framework/src/utils/InputParameters.C

```diff
@@ -1195,3 +1230,61 @@ InputParameters::varName(const std::string & var_param_name,
 
   return variable_name;
 }
+
+void
+InputParameters::renameParam(const std::string & old_name,
+                             const std::string & new_name,
+                             const std::string & new_docstring)
+{
+  auto params_it = _params.find(old_name);
+  if (params_it == _params.end())
+    mooseError("Requested to rename parameter '",
+               old_name,
+               "' but that parameter name doesn't exist in the parameters object.");
+
+  auto new_metadata = std::move(params_it->second);
+  new_metadata._doc_string = new_docstring;
+  _params.emplace(new_name, std::move(new_metadata));
+  _params.erase(params_it);
+
+  auto values_it = _values.find(old_name);
+  auto new_value = std::move(values_it->second);
+  _values.emplace(new_name, std::move(new_value));
+  _values.erase(values_it);
+
+  _old_to_new_name.emplace(old_name, new_name);
+  // invalidate the cache
+  _last_checked_rename.first.clear();
+}
+
+void
+InputParameters::renameCoupledVar(const std::string & old_name,
+                                  const std::string & new_name,
+                                  const std::string & new_docstring)
+{
+  auto coupled_vars_it = _coupled_vars.find(old_name);
+  if (coupled_vars_it == _coupled_vars.end())
+    mooseError("Requested to rename coupled variable '",
+               old_name,
+               "' but that coupled variable name doesn't exist in the parameters object.");
+
+  _coupled_vars.insert(new_name);
+  _coupled_vars.erase(coupled_vars_it);
+
+  renameParam(old_name, new_name, new_docstring);
+}
+
+std::string
+InputParameters::checkForRename(const std::string & name_in) const
+{
+  if (name_in == _last_checked_rename.first)
+    return _last_checked_rename.second;
+
+  _last_checked_rename.first = name_in;
+  if (auto it = _old_to_new_name.find(name_in); it != _old_to_new_name.end())
+    _last_checked_rename.second = it->second;
```

Comment:
Do we issue a deprecation message here?

Reply:
Is this for deprecation or for derived classes to provide more meaningful parameter names?

Reply:
The latter

Reply:
Would it be useful for deprecation, if we added a flag to distinguish the two use cases and avoid false positive deprecation messages?  Deprecating poor initial choices of parameter names was the first use case that came to my mind too.

Reply:
I can think about how to integrate this into our existing deprecation APIs